src/ripple/basics/impl/Number.cpp

```diff
@@ -0,0 +1,732 @@
+            xe = ye;
+            xn = yn;
+        }
+        while (xm < minMantissa)
+        {
+            xm *= 10;
+            xm -= g.pop();
+            --xe;
+        }
+        auto r = g.round();
+        if (r == 1 || (r == 0 && (xm & 1) == 1))
+        {
+            --xm;
+            if (xm < minMantissa)
+            {
+                xm *= 10;
+                --xe;
+            }
+        }
+        if (xe < minExponent)
+        {
+            xm = 0;
+            xe = Number{}.exponent_;
+        }
+    }
+    mantissa_ = xm * xn;
+    exponent_ = xe;
+    return *this;
+}
+
+Number&
+Number::operator*=(Number const& y)
+{
+    if (*this == Number{})
+        return *this;
+    if (y == Number{})
+    {
+        *this = y;
+        return *this;
+    }
+    assert(isnormal() && y.isnormal());
+    auto xm = mantissa();
+    auto xe = exponent();
+    int xn = 1;
+    if (xm < 0)
+    {
+        xm = -xm;
+        xn = -1;
+    }
+    auto ym = y.mantissa();
+    auto ye = y.exponent();
+    int yn = 1;
+    if (ym < 0)
+    {
+        ym = -ym;
+        yn = -1;
+    }
+    auto zm = uint128_t(xm) * uint128_t(ym);
+    auto ze = xe + ye;
+    auto zn = xn * yn;
+    Guard g;
+    while (zm > maxMantissa)
+    {
+        g.push(static_cast<unsigned>(zm % 10));
+        zm /= 10;
+        ++ze;
+    }
+    xm = static_cast<rep>(zm);
+    xe = ze;
+    auto r = g.round();
+    if (r == 1 || (r == 0 && (xm & 1) == 1))
+    {
+        ++xm;
+        if (xm > maxMantissa)
+        {
+            xm /= 10;
+            ++xe;
+        }
+    }
+    if (xe < minExponent)
+    {
+        xm = 0;
+        xe = Number{}.exponent_;
+    }
+    if (xe > maxExponent)
+        throw std::overflow_error(
+            "Number::multiplication overflow : exponent is " +
+            std::to_string(xe));
+    mantissa_ = xm * zn;
+    exponent_ = xe;
+    assert(isnormal() || *this == Number{});
+    return *this;
+}
+
+Number&
+Number::operator/=(Number const& y)
+{
+    if (y == Number{})
+        throw std::overflow_error("Number: divide by 0");
+    int np = 1;
+    auto nm = mantissa();
+    if (nm < 0)
+    {
+        nm = -nm;
+        np = -1;
+    }
+    int dp = 1;
+    auto dm = y.mantissa();
+    if (dm < 0)
+    {
+        dm = -dm;
+        dp = -1;
+    }
+    // Divide numerator and denominator such that the
+    //     denominator is in the range [1, 10).
+    const int offset = -15 - y.exponent();
+    Number n{nm * (np * dp), exponent() + offset};
+    Number d{dm, y.exponent() + offset};
+    // Quadratic least squares fit to 1/x in the range [1, 10]
+    constexpr Number a0{9178756872006464, -16, unchecked{}};
+    constexpr Number a1{-2149215784206187, -16, unchecked{}};
+    constexpr Number a2{1405502114116773, -17, unchecked{}};
+    static_assert(a0.isnormal());
+    static_assert(a1.isnormal());
+    static_assert(a2.isnormal());
+    Number rm2{};
+    Number rm1{};
+    Number r = (a2 * d + a1) * d + a0;
```

Comment:
I understand why you're doing this, but I don't love that division is not an iterative process. This can't be good for performance. No change needed, just making a note.

Reply:
Taken from [Fast division methods](https://en.wikipedia.org/wiki/Division_algorithm#Fast_division_methods).